submission/races_test.go

```diff
@@ -0,0 +1,152 @@
+	t.Helper()
+	for _, sct := range got {
+		if _, ok := trail[ctpolicy.BaseName]; ok {
+			trail[ctpolicy.BaseName]--
+			if trail[sct.logURL[0:1]] > 0 {
+				trail[sct.logURL[0:1]]--
+			}
+		} else {
+			trail[sct.logURL[0:1]]--
+		}
+	}
+	for groupName, count := range trail {
+		if count < 0 {
+			t.Errorf("Received %d more SCTs from group %q than expected", count*(-1), groupName)
+		} else if count > 0 {
+			t.Errorf("Received %d less SCTs from group %q than expected", count, groupName)
+		}
+	}
+}
+
+func TestGetSCTs(t *testing.T) {
+	testCases := []struct {
+		name        string
+		sbMock      Submitter
+		groups      ctpolicy.LogPolicyData
+		ctx         context.Context
+		resultTrail map[string]int
+		errRegexp   *regexp.Regexp
+	}{
+		{
+			name:   "singleGroupOneSCT",
+			sbMock: &mockSubmitter{firstLetterURLReqNumber: make(map[byte]int)},
+			groups: ctpolicy.LogPolicyData{
+				"a": {
+					Name:          "a",
+					LogURLs:       map[string]bool{"a1.com": true, "a2.com": true},
+					MinInclusions: 1,
+					IsBase:        false,
+				},
+			},
+			ctx:         context.Background(),
+			resultTrail: map[string]int{"a": 1},
+		},
+		{
+			name:   "singleGroupMultiSCT",
+			sbMock: &mockSubmitter{firstLetterURLReqNumber: make(map[byte]int)},
+			groups: ctpolicy.LogPolicyData{
+				"a": {
+					Name:          "a",
+					LogURLs:       map[string]bool{"a1.com": true, "a2.com": true, "a3.com": true, "a4.com": true, "a5.com": true},
+					MinInclusions: 3,
+					IsBase:        false,
+				},
+			},
+			ctx:         context.Background(),
+			resultTrail: map[string]int{"a": 3},
+		},
+		{
+			name:   "chromeLike",
+			sbMock: &mockSubmitter{firstLetterURLReqNumber: make(map[byte]int)},
+			groups: ctpolicy.LogPolicyData{
+				"a": {
+					Name:          "a",
+					LogURLs:       map[string]bool{"a1.com": true, "a2.com": true, "a3.com": true, "a4.com": true},
+					MinInclusions: 1,
+					IsBase:        false,
+				},
+				"b": {
+					Name:          "b",
+					LogURLs:       map[string]bool{"b1.com": true, "b2.com": true, "b3.com": true, "b4.com": true},
+					MinInclusions: 1,
+					IsBase:        false,
+				},
+				"Base": {
+					Name:          "Base",
+					LogURLs:       map[string]bool{"a1.com": true, "a2.com": true, "a3.com": true, "a4.com": true, "b1.com": true, "b2.com": true, "b3.com": true, "b4.com": true},
+					MinInclusions: 3,
+					IsBase:        true,
+				},
+			},
+			ctx:         context.Background(),
+			resultTrail: map[string]int{"a": 1, "b": 1, ctpolicy.BaseName: 3},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			res, err := GetSCTs(tc.ctx, tc.sbMock, []ct.ASN1Cert{{Data: []byte{0}}}, tc.groups)
+			if tc.resultTrail != nil {
+				evaluateSCTs(t, res, tc.resultTrail)
+			}
+			if tc.errRegexp != nil {
+				if !tc.errRegexp.MatchString(err.Error()) {
+					t.Errorf("Error %q did not match expected regexp %q", err, tc.errRegexp)
+				}
+			}
+		})
+	}
+}
```

Comment:
Could do with a bunch more test cases here, but probably worth holding off until we're settled on the mainline code.  

Things to possibly test:
 - `safeStringSet` behaviour
 - individual methods like `postInterval()`, `parallelNums()`, `groupsComplete()`
 - more variations of submission failures, timeouts, log counts etc (with the aim of hitting every possible exit condition from `groupRace` and from the loop in `GetSCTs`)

Reply:
I thought the comment on the very beginning of this review reflects why there's not that many tests (yet).

Reply:
Ah yes, I forgot about that.